sdk/src/action.ts

```diff
@@ -90,11 +95,21 @@ export class Action<
 
     this.grpc = grpc.loadPackageDefinition(packageDefinition);
 
-    this.clients = GrpcUtil.clientCreators(this.root, this.grpc);
+    this.clients = undefined;
 
     this.commandHandlers = {} as CommandHandlers;
   }
 
+  preStart(settings: PreStartSettings): void {
+    this.clients = GrpcUtil.clientCreators(
+      this.root,
+      this.grpc,
+      settings.proxyHostname,
+      settings.proxyPort,
+      settings.identificationInfo,
+    );
```

Comment:
Not setting this until discovery breaks the current samples which creates a client immediately on start.

Reply:
Still went with this breaking change for now, couldn't come up with anything sensible to allow create early but later injection. Perhaps we can improve in the future.

Reply:
> Still went with this breaking change for now, couldn't come up with anything sensible to allow create early but later injection. Perhaps we can improve in the future.

Makes sense. The components are created directly. If it was a class-based approach for defining the component behaviour, then could be easier? But that would be a bigger breaking change again.

Reply:
One thing we could perhaps do is to pass the clients via the various `context`s instead of as a field on the component object itself, that way it is obvious when it is available only when there is a context.

Reply:
👍🏼 Having it on the `context` could make more sense.